app/models/smart_proxy.rb

```diff
@@ -172,6 +172,6 @@ def get_features
   end
 
   class Jail < ::Safemode::Jail
-    allow :id, :name
+    allow :id, :name, :hostname, :hosts_count, :setting
```

Comment:
Does `setting` allow to retrieve any secret settings?

Reply:
It allows retrieving settings of a smart proxy feature. I don't think those contain any secrets

Reply:
Should we perhaps expose those explicitly? Features are just strings and exposed on the individual proxies without any authentication so that's IMHO safe. If you're talking about exposed settings from a proxy then those can contain private information and they're only available via an authenticated endpoint (though I'm not aware of any secrets being passed now).

Reply:
From my testing machine:
```
SmartProxyFeature.all.each { |f| puts "#{f.feature.name}: #{f.settings}" }
Pulp: {"pulp_url"=>"https://centos7-katello-nightly-stable.example.com/pulp"}
Pulpcore: {"pulp_url"=>"https://centos7-katello-nightly-stable.example.com", "mirror"=>false, "content_app_url"=>"https://centos7-katello-nightly-stable.example.com/pulp/content"}
TFTP: {"tftp_servername"=>nil}
Puppet CA: {"puppet_url"=>"https://centos7-katello-nightly-stable.example.com:8140", "use_provider"=>["puppetca_hostname_whitelisting", "puppetca_http_api"]}
Puppet: {"puppet_url"=>"https://centos7-katello-nightly-stable.example.com:8140", "use_provider"=>["puppet_proxy_puppet_api"]}
Logs: {}
Templates: {}
HTTPBoot: {"http_port"=>8000, "https_port"=>9090}
```

Reply:
https://github.com/theforeman/smart_proxy_pulp/blob/55ff29f4cfab971f71bc00273c7623ee8dc5af96/lib/smart_proxy_pulp_plugin/pulpcore_plugin.rb#L17-L18 does expose a username/password. It's mostly used in the development workflow if you don't have certificates set up so you won't see it in production, but exposing this is a security risk.

Another thing to consider is that these settings are IMHO internal implementation details. You should always wrap them in high level logic. For example, the puppet server url has a one that also provides compatibility with older releases which didn't expose this:

https://github.com/theforeman/foreman/blob/c3560417748c744d6f64336b1b98102e6a8bb4c1/app/models/concerns/host_common.rb#L95-L100

That particular helper should be exposed rather than raw settings.

:-1: on exposing `setting`.

Reply:
Ok how about that, I have implemented whitelisting.